File: public/src/@types/auth/response.ts
// success
export type SuccessLoginAPI = {
    message: string;
    data: {
        access_token: string;
    },
}

export type SuccessRegisterAPI = {
    message: string;
    data: {
        id: string;
        email: string;
        is_verified: boolean;
        role_id: number;
        role: string;
        created_at: string;
    },
}

export interface SuccessGetMeAPI {
    message: string;
    data: UserData;
}

export interface UserDetail {
    id: number;
    user_id: number;
    full_name: string;
    sipa_number: string;
    whatsapp_number: string;
    years_of_experience: number;
    image_url: string;
    created_at: string;
    updated_at: string;
    deleted_at: string | null;
}

export interface UserData {
    id: number;
    role_id: number;
    role: string;
    email: string;
    is_verified: boolean;
    user_detail: UserDetail | null
    address: Address[];
    created_at: string;
    updated_at: string;
    deleted_at: string | null;
}

export interface Address {
    id: number;
    userId: number;
    is_active: boolean;
    address: string;
    province: string;
    city: string;
    city_id: number
    district: string;
    sub_district: string;
    latitude: string;
    longitude: string;
    createdAt: string;
    updatedAt: string;
    deletedAt?: string | null;
}

export interface JwtPayload {
    iss: string;
    exp: number;
    iat: number;
    jti: string;
    user_id: number;
    role: number;
}

File: server/internal/profile/dto/response.go
package dto

import (
	"strings"
	"time"

	dtoAuth "healthcare-app/internal/auth/dto"
	"healthcare-app/internal/profile/entity"
)

type ResponseAddress struct {
	ID                 int64      `json:"id"`
	UserId             int64      `json:"user_id"`
	IsActive           bool       `json:"is_active"`
	Address            string     `json:"address"`
	Province           string     `json:"province"`
	CityID             int64      `json:"city_id"`
	City               string     `json:"city"`
	District           string     `json:"district"`
	SubDistrict        string     `json:"sub_district"`
	Latitude           string     `json:"latitude"`
	Longitude          string     `json:"longitude"`
	ContactName        string     `json:"contact_name"`
	ContactPhoneNumber string     `json:"contact_phone_number"`
	CreatedAt          time.Time  `json:"created_at"`
	UpdatedAt          time.Time  `json:"updated_at"`
	DeletedAt          *time.Time `json:"deleted_at,omitempty"`
}

type ResponseProfile struct {
	ID         int64                       `json:"id"`
	RoleId     int                         `json:"role_id"`
	Role       string                      `json:"role"`
	Email      string                      `json:"email"`
	IsVerified bool                        `json:"is_verified"`
	UserDetail *dtoAuth.ResponseUserDetail `json:"user_detail"`
	Address    []*ResponseAddress          `json:"address"`
	CreatedAt  time.Time                   `json:"created_at"`
	UpdatedAt  time.Time                   `json:"updated_at"`
	DeletedAt  *time.Time                  `json:"deleted_at"`
}

func ConvertToAddressResponses(addresses []*entity.Address) []*ResponseAddress {
	responses := []*ResponseAddress{}
	for _, address := range addresses {
		responses = append(responses, ConvertToAddressResponse(address))
	}
	return responses
}

func ConvertToAddressResponse(address *entity.Address) *ResponseAddress {
	location := strings.Split(address.Location, " ")

	return &ResponseAddress{
		ID:                 address.ID,
		UserId:             address.UserId,
		IsActive:           address.IsActive,
		Address:            address.Address,
		Province:           address.Province,
		CityID:             address.CityID,
		City:               address.City,
		District:           address.District,
		SubDistrict:        address.SubDistrict,
		Longitude:          location[0],
		Latitude:           location[1],
		ContactName:        address.ContactName,
		ContactPhoneNumber: address.ContactPhoneNumber,
		CreatedAt:          address.CreatedAt,
		UpdatedAt:          address.UpdatedAt,
	}
}
